app/features/connect_with_them/tools.py

```diff
@@ -0,0 +1,137 @@
+import requests
+from requests.auth import HTTPBasicAuth
+from requests.exceptions import RequestException
+from langchain import hub
+from langchain.agents import AgentExecutor,create_react_agent,tool
+from langchain_community.document_loaders import WebBaseLoader
+from langchain.chains.summarize.chain import load_summarize_chain
+from langchain.tools.base import StructuredTool
+from langchain_groq import ChatGroq
+from langchain.pydantic_v1 import BaseModel, Field
+
+from app.features.connect_with_them.Credentials import credentials,categories_mapping
+from app.features.connect_with_them.prompt.Prompts import Prompt_query
+
+from bs4 import BeautifulSoup
+
+model_name = 'llama-3.1-70b-versatile'
+
+llm = ChatGroq(model=model_name,temperature=0.3,api_key="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")
+
+class GetSubredditsModel(BaseModel):
+    Topic: str = Field(description='Topic that you want to get subreddits for')
+
+class SearchArticlesModel(BaseModel):
+    category: str = Field(description='Category that you want to get some information about')
+
+def get_subreddits(topic:str ,limit:int=10 ) -> str:
+    params = {'q': topic, 'type': 'sr', 'sort': 'relevance', 'limit': limit}
+    data = {'grant_type': 'password', 'username': credentials['NAME'], 'password': credentials['PASSWORD']}
+    headers = {'User-Agent': credentials['USER_AGENT']}
+    try:
+        auth = HTTPBasicAuth(credentials['CLIENT_ID'], credentials['CLIENT_SECRET'])
+        res = requests.post('https://www.reddit.com/api/v1/access_token', auth=auth, data=data, headers=headers)
+        res.raise_for_status()
+        token = res.json()['access_token']
+        headers['Authorization'] = f'bearer {token}'
+        response = requests.get('https://oauth.reddit.com/subreddits/search', headers=headers, params=params)
+        response.raise_for_status()
+        data = response.json()['data']['children']
+        subreddits = [post['data']['display_name'] for post in data]
+        return f"These are the subreddits related to '{topic}': {', '.join(subreddits)}"
+    except RequestException as e:
+        print(f"An error occurred during the request: {e}")
+        return f"No subreddits found for '{topic}'. Error: {str(e)}"
+
+def scrape_page(category: str, num_articles: int) -> list:
+    # Build the URL based on the category
+    url = f"https://www.edutopia.org/{categories_mapping[category]}"
+
+    try:
+        # Send a GET request to fetch the page content
+        response = requests.get(url)
+        response.raise_for_status()  # Raise an exception for HTTP errors
+
+        # Parse the HTML content using BeautifulSoup
+        soup = BeautifulSoup(response.text, 'html.parser')
```

Comment:
Don't scrape the pages since we need document loaders and it'll be complicated to integrate with the required info.